clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java

```diff
@@ -1207,6 +1232,73 @@ public void run() {
     }
   }
 
+  /**
+   * Get the set of regions that haven't been pushed yet for targeted region push.
+   * @return a set of regions that haven't been pushed yet.
+   */
+  private Set<String> getRegionsForTargetedRegionPush() {
+    // only kick off the validation and post-validation flow when everything has to be done in a single VPJ
+    if (pushJobSetting.isTargetedRegionPushEnabled && pushJobSetting.postValidationConsumption) {
+      Set<String> targetedRegions = RegionUtils.parseRegionsFilterList(pushJobSetting.targetedRegions);
+      Set<String> candidateRegions =
+          new HashSet<>(storeSetting.storeResponse.getStore().getColoToCurrentVersions().keySet());
+      candidateRegions.removeAll(targetedRegions);
+      return candidateRegions;
+    }
+    return Collections.emptySet();
+  }
+
+  /**
+   * Start a group of validation job to verify result from the previous targeted region push is healthy.
+   */
+  void postPushValidation() {
+    if (kafkaTopicInfo.kafkaSourceRegion == null) {
+      throw new VeniceException("Post-validation consumption halted due to no available source region found");
+    }
+
+    // TODO: Add parallelism support when we have more validators. It's premature to do so now.
+    Validator noOpValidator = new NoOpValidator();
+    noOpValidator.validate();
+  }
+
+  /**
+   * Kick off the post-validation consumption on the rest of regions that haven't been pushed yet.
+   * @param candidateRegions
+   */
+  private void postValidationConsumption(Set<String> candidateRegions) {
+    // get the latest version of the store for data recovery
+    Version sourceVersion = getStoreVersion(pushJobSetting.storeName, kafkaTopicInfo.version);
+
+    if (sourceVersion.getHybridStoreConfig() != null) {
+      // perform data recovery for HYBRID stores
+      pushJobSetting.isTargetedRegionPushEnabled = false;
+      pushJobSetting.targetedRegions = null;
+      // set up repush
+      pushJobSetting.isSourceKafka = true;
+      pushJobSetting.kafkaInputBrokerUrl = kafkaTopicInfo.kafkaUrl;
+      pushJobSetting.kafkaInputTopic = kafkaTopicInfo.topic;
+      this.run();
+    } else {
+      // perform data recovery for BATCH stores
+      for (String region: candidateRegions) {
+        ControllerResponse response = controllerClient.dataRecovery(
+            kafkaTopicInfo.kafkaSourceRegion,
+            region,
+            pushJobSetting.storeName,
+            sourceVersion.getNumber(),
+            true,
+            true,
+            Optional.of(sourceVersion));
+        if (response.isError()) {
+          throw new VeniceException("Can't perform data recovery for region " + region + ". " + response.getError());
+        }
+      }
+      // Essentially, the remaining regions are "targeted" regions now
+      pushJobSetting.targetedRegions = RegionUtils.composeRegionList(candidateRegions);
+      pollStatusUntilComplete(Optional.empty(), controllerClient, pushJobSetting, kafkaTopicInfo);
```

Comment:
I was thinking is it cleaner to pass target regions to the method `pollStatusUntilComplete` instead of hacking it outside ?

Reply:
Sure.